Update client code for S22

NicMcPhee opened the pull request:
Bring over some other changes (other than all the dependency changes in #34) from Lab 3 and the iteration template that seem to make sense here.

This includes:

* Fixing the titles to be Lab 4 instead of the iteration template
* Using the sidenav Cypress tests from Lab 3 since they're a lot nicer
* Bringing a bunch of comments (and some restructuring) forward from Lab 3

I'm pretty sure I've duplicated a few changes that are in #34, so we should merge that in first and then see how that affects this.

NicMcPhee committed:
Replace all "Iteration Template" titles with "Lab 4"

NicMcPhee committed:
Replace the sidenav Cypress description with the one from Lab 3

Lab 3's Cypress test for the sidenav in `app.spec.ts` makes a lot more sense than the test in either Lab 4 or the iteration template, so I'm copying it over here.

NicMcPhee committed:
Remove unnecessary constructor for `UserCardComponent`

NicMcPhee committed:
Clean up a few issues in `UserListComponent`

NicMcPhee committed:
Bring forward the changes to `UserListComponent` spec from Lab 3

I made several changes to the Karma tests for `UserListComponent` in Lab 3 that this brings forward. It also required bringing forward a change to `.eslintrc.json` so that `ng lint` doesn't fuss about the `constructUserList()` helper function.

NicMcPhee committed:
Rationalize UserListComponent

This component is different in each of Lab 3, Lab 4, and the iteration template. This tries to bring together the "best" bits from each of those.

I'm still not thrilled with the `unsubscribe()` business, but I'm not sure what a better approach would be.

NicMcPhee committed:
Migrate `UserService` specs from Lab 3

Lab 3 has some nice comments and restructuring that isn't in Lab 4 or the iteration template, so this migrates that forward to Lab 4.

NicMcPhee committed:
Migrate forward Lab 3 `UserService` comments

This brings a bunch of comments I added to `UserService` in Lab 3 to also be in here in Lab 4.

NicMcPhee committed:
Migrate back an iteration template comment in `app-routing.module.ts`

This comment from the iteration template looks useful so I migrated it back to this lab.

NicMcPhee committed:
Tidy up the formatting on some of the `app` files

NicMcPhee committed:
Migrate forward some Lab 3 comments from on mock user service

NicMcPhee committed:
Delete some unnecessary client config elements

helloworld12321 commented on client/src/app/users/user-list.component.spec.ts:
```diff
@@ -83,10 +108,17 @@ describe('User list', () => {
   });
 });
 
+/*
+ * This test is a little odd, but illustrates how we can use stubs
+ * to create mock objects (a service in this case) that be used for
+ * testing. Here we set up the mock UserService (userServiceStub) so that
+ * _always_ fails (throws an exception) when you request a set of users.
+ *
+ * So this doesn't really test anything meaningful in the context of our
+ * code (I certainly wouldn't copy it), but it does illustrate some nice
+ * testing tools. Hopefully it's useful as an example in that regard.
+ */
```
Oh, I'm not sure if I agree with that. This test case _does_ test something meaningful—namely, what happens to the component if we can't talk to the server? It's useful to know that `userList.serverFilteredUsers` is undefined if that happens, rather than having some garbage value.

NicMcPhee replied:
I'll buy that, and just removed that second paragraph.

helloworld12321 reviewed (commented):
Looks great, just one comment, see below. 🙂 I like how you added a bunch of notes to explain the test code :)

NicMcPhee committed:
Remove overly negative comment in UserListComponent spec

@helloworld12321 pointed out that this comment isn't really correct since something useful _is_ being tested here. The test could probably be improved, but it's still better than nothing.

helloworld12321 reviewed (approved):
Looks excellent, thanks so much!
